software/controller/controller.sh

```diff
@@ -100,85 +148,117 @@ run_integration_tests() {
+  lcov -r "$COVERAGE_OUTPUT_DIR/$COVERAGE_ENVIRONMENT.info" \
+      --output-file "$COVERAGE_OUTPUT_DIR/${COVERAGE_ENVIRONMENT}_trimmed.info" \
+      "*output_export.c*" \
+      "*.pio/libdeps/*" \
+      "*_test_transport.c" \
+      "/usr/include*"
+  genhtml "$COVERAGE_OUTPUT_DIR/${COVERAGE_ENVIRONMENT}_trimmed.info" \
+      --output-directory "$COVERAGE_OUTPUT_DIR"
+
+  echo "Coverage report generated. Open '$COVERAGE_OUTPUT_DIR/index.html' in a browser to view it."
+}
+
+launch_browser() {
+  python -m webbrowser ${COVERAGE_OUTPUT_DIR}/index.html
+}
+
 ########
 # HELP #
 ########
 
-if [ "$1" == "--help" ]; then
+if [ "$1" == "--help" ] || [ "$1" == "-h" ]; then
   print_help
   exit $EXIT_SUCCESS
-fi
 
 ###########
 # INSTALL #
 ###########
-
-if [ "$1" == "--install" ]; then
+elif [ "$1" == "install" ]; then
   if [ "$EUID" -eq 0 ] && [ "$2" != "-f" ]; then
     echo "Please do not run install with root privileges!"
     exit $EXIT_SUCCESS
   fi
 
-  pushd "$HOME"
-  python3 -c "$(curl -fsSL https://raw.githubusercontent.com/platformio/platformio/master/scripts/get-platformio.py)"
-  echo 'export PATH=$PATH:~/.platformio/penv/bin' >> ~/.profile
-  popd
-
-  curl -fsSL https://raw.githubusercontent.com/platformio/platformio-core/master/scripts/99-platformio-udev.rules | sudo tee /etc/udev/rules.d/99-platformio-udev.rules
-  sudo service udev restart
-  sudo usermod -a -G dialout "$USER"
-  sudo usermod -a -G plugdev "$USER"
-
-  echo "Updated udev rules. You might have to restart your machine for changes to become effective."
-
-  exit $EXIT_SUCCESS
-fi
+  configure_platformio
 
 #########
 # CLEAN #
 #########
-
-if [ "$1" == "--clean" ]; then
-  clean_dir .pio
+elif [ "$1" == "clean" ]; then
+  clean_all
   exit $EXIT_SUCCESS
-fi
 
 #########
 # CHECK #
 #########
-
-if [ "$1" == "--check" ]; then
+elif [ "$1" == "check" ]; then
   run_checks
   exit $EXIT_SUCCESS
-fi
 
-########
-# TEST #
-########
+#############
+# UNIT TEST #
+#############
+elif [ "$1" == "unit" ]; then
+  clean_all
 
-if [ "$1" == "--test" ]; then
-  # Controller unit tests on native.
-  pio test -e native
+  if [ -n "$2" ]; then
+    pio test -e native -f "$2"
+  else
+    pio test -e native
+  fi
 
-  run_integration_tests
+  generate_coverage_reports
 
-  # Make sure controller builds for target platform.
+  exit $EXIT_SUCCESS
+
+############
+# TEST ALL #
+############
+elif [ "$1" == "test" ]; then
+  clean_all
+
+  # Make sure controller builds for target platform
   pio run -e stm32
 
-  if [ "$2" != "--no-checks" ]; then
+  # Make sure integration tests build
+  run_integration_tests
+
+  # Controller unit tests on native
+  # This must be the last thing built
+  clean_all
+  pio test -e native
+
+  if [ "$2" != "--no-checks" ] && [ "$3" != "--no-checks" ]; then
     run_checks
   else
     echo "Skipping static checks."
   fi
 
+  if [ "$2" != "--no-cov" ] && [ "$3" != "--no-cov" ]; then
```

Comment:
They are actually pretty fast locally.
My experience has been, when I write new tests, that it is useful to do this locally instead of idly waiting for CI builds to finish. In CI, the entire thing has to finish before it will publish any reports. But now you can do, for example `./controller.sh test debug` and you will get a report for specifically what your module is testing way faster.